src/buffer.rs

```diff
@@ -1580,7 +1618,14 @@ impl UnicodeBuffer {
     /// Pushes a string to a buffer.
     #[inline]
     pub fn push_str(&mut self, str: &str) {
-        self.0.push_str(str);
+        self.0.add_utf8(str, 0, str.len());
+    }
+
+    /// Pushes a substring of a larger string to a buffer.
+    /// The rest of the string acts as context for shaping.
+    #[inline]
+    pub fn add_utf8(&mut self, str: &str, item_offset: usize, item_length: usize) {
```

Comment:
Shouldn't we simply accept 3 `&str`?

Reply:
Good point. The signature above follows HarfBuzz’s API, but I think accepting 3 `&str`s is more elegant as a solution. It at least removes the possibility of panicking if `item_offset` and `item_offset + item_length` don’t lie on UTF-8 boundaries.

Reply:
Come to think of it, we could alternatively add separate methods for adding pre- and post-context. I’m not sure which design is better.

Reply:
Yes, `set_pre_context` and `set_post_context` is a better choice.